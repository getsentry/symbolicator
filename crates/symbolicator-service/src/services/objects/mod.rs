use std::collections::BTreeSet;
use std::sync::Arc;

use futures::future;
use sentry::{Hub, SentryFutureExt};

use symbolicator_sources::{FileType, ObjectId, RemoteFile, RemoteFileUri, SourceConfig, SourceId};

use crate::caching::{Cache, CacheEntry, CacheError, CacheKey, Cacher, SharedCacheRef};
use crate::services::download::DownloadService;
use crate::types::{AllObjectCandidates, ObjectCandidate, ObjectDownloadInfo, Scope};

use data_cache::FetchFileDataRequest;
use meta_cache::FetchFileMetaRequest;

pub use data_cache::ObjectHandle;
pub use meta_cache::ObjectMetaHandle;

mod data_cache;
mod meta_cache;

/// Wrapper around [`CacheError`] to also pass the file information along.
///
/// Because of the requirement of [`CacheItemRequest`] to impl `From<io::Error>` it can not
/// itself contain the [`SourceId`] and [`SourceLocation`].  However we need to carry this
/// along some errors, so we use this wrapper.
///
/// [`CacheItemRequest`]: crate::caching::CacheItemRequest
/// [`SourceLocation`]: crate::services::download::SourceLocation
#[derive(Clone, Debug)]
pub struct CacheLookupError {
    /// The object file which was attempted to be fetched.
    pub file_source: RemoteFile,
    /// The wrapped [`CacheError`] which occurred while fetching the object file.
    pub error: CacheError,
}

/// Fetch a Object from external sources or internal cache.
#[derive(Debug, Clone)]
pub struct FindObject {
    pub filetypes: &'static [FileType],
    pub purpose: ObjectPurpose,
    pub scope: Scope,
    pub identifier: ObjectId,
    pub sources: Arc<[SourceConfig]>,
}

#[derive(Debug, Copy, Clone)]
pub enum ObjectPurpose {
    Unwind,
    Debug,
    Source,
}

#[derive(Debug, Clone)]
pub struct FoundMeta {
    pub file_source: RemoteFile,
    pub handle: CacheEntry<Arc<ObjectMetaHandle>>,
}

/// The response for [`ObjectsActor::find`].
///
/// The found object is in the `selected_meta` field with other DIFs considered
/// in the `candidates` field.
#[derive(Debug, Clone)]
pub struct FindResult {
    /// If a matching object (or an error as fallback) was found,
    /// its [`ObjectMetaHandle`] will be provided here, otherwise this will be `None`
    pub meta: Option<FoundMeta>,
    /// This is a list of some meta information on all objects which have been considered
    /// for this object.  It could be populated even if no matching object is found.
    pub candidates: AllObjectCandidates,
}

#[derive(Clone, Debug)]
pub struct ObjectsActor {
    meta_cache: Arc<Cacher<FetchFileMetaRequest>>,
    data_cache: Arc<Cacher<FetchFileDataRequest>>,
    download_svc: Arc<DownloadService>,
}

impl ObjectsActor {
    pub fn new(
        meta_cache: Cache,
        data_cache: Cache,
        shared_cache: SharedCacheRef,
        download_svc: Arc<DownloadService>,
    ) -> Self {
        ObjectsActor {
            meta_cache: Arc::new(Cacher::new(meta_cache, Arc::clone(&shared_cache))),
            data_cache: Arc::new(Cacher::new(data_cache, shared_cache)),
            download_svc,
        }
    }

    /// Returns the requested object file.
    ///
    /// This fetches the requested object, re-downloading it from the source if it is no
    /// longer in the cache.
    pub async fn fetch(&self, file_handle: Arc<ObjectMetaHandle>) -> CacheEntry<Arc<ObjectHandle>> {
        let cache_key = CacheKey::from_scoped_file(&file_handle.scope, &file_handle.file_source);
        let request = FetchFileDataRequest(FetchFileMetaRequest {
            scope: file_handle.scope.clone(),
            file_source: file_handle.file_source.clone(),
            object_id: file_handle.object_id.clone(),
            data_cache: self.data_cache.clone(),
            download_svc: self.download_svc.clone(),
        });

        self.data_cache.compute_memoized(request, cache_key).await
    }

    /// Fetches matching objects and returns the metadata of the most suitable object.
    ///
    /// This requests the available matching objects from the sources and then looks up the
    /// object metadata of each matching object in the metadata cache.  These are then
    /// ranked and the best matching object metadata is returned.
    ///
    /// Asking for the objects metadata from the data cache also triggers a download of each
    /// object, which will then be cached in the data cache.  The metadata itself is cached
    /// in the metadata cache which usually lives longer.
    pub async fn find(&self, request: FindObject) -> FindResult {
        let FindObject {
            filetypes,
            scope,
            identifier,
            sources,
            purpose,
        } = request;
        let file_ids = self
            .download_svc
            .list_files(&sources, filetypes, &identifier)
            .await;

        let file_metas = self.fetch_file_metas(file_ids, &identifier, scope).await;

        let candidates = create_candidates(&sources, &file_metas);
        let meta = select_meta(file_metas, purpose);

        FindResult { meta, candidates }
    }

    /// Fetch all [`ObjectMetaHandle`]s for the files.
    ///
    /// This concurrently looks up the file IDs in the meta-cache and returns all results.
    /// A custom [`CacheLookupError`] is returned to allow us to keep track of the source ID
    /// and source location in case of an error.  [`select_meta`] uses this to build the
    /// [`ObjectCandidate`] list.
    async fn fetch_file_metas(
        &self,
        file_sources: Vec<RemoteFile>,
        identifier: &ObjectId,
        scope: Scope,
    ) -> Vec<FoundMeta> {
        let queries = file_sources.into_iter().map(|file_source| {
            let scope = if file_source.is_public() {
                Scope::Global
            } else {
                scope.clone()
            };
            let cache_key = CacheKey::from_scoped_file(&scope, &file_source);
            let request = FetchFileMetaRequest {
                scope,
                file_source: file_source.clone(),
                object_id: identifier.clone(),
                data_cache: self.data_cache.clone(),
                download_svc: self.download_svc.clone(),
            };

            let meta_cache = self.meta_cache.clone();

            async move {
                let handle = meta_cache.compute_memoized(request, cache_key).await;
                FoundMeta {
                    file_source,
                    handle,
                }
            }
            .bind_hub(Hub::new_from_top(Hub::current()))
        });

        future::join_all(queries).await
    }
}

/// Select the best [ObjectMetaHandle`] out of all lookups from the meta-cache.
///
/// The lookups are expected to be in order or preference, so if two files are equally good
/// the first one will be chosen.  If the file list is emtpy, `None` is returned in the
/// result, if there were no suitable files and only lookup errors one of the lookup errors
/// is propagated.  If there were no suitlable files and no errors `None` is also returned
/// in the result.
fn select_meta(all_lookups: Vec<FoundMeta>, purpose: ObjectPurpose) -> Option<FoundMeta> {
    let mut selected_meta = None;
    let mut selected_quality = u8::MAX;

    for meta_lookup in all_lookups {
        // Skip objects which and not suitable for what we're asked to provide.  Keep errors
        // though, if we don't find any object we need to return an error.
        if let Ok(ref meta_handle) = meta_lookup.handle {
            if !object_has_features(meta_handle, purpose) {
                continue;
            }
        }

        // We iterate in order of preferred sources, so only select a later object if the
        // quality is better.
        let quality = object_quality(&meta_lookup, purpose);
        if quality < selected_quality {
            selected_meta = Some(meta_lookup);
            selected_quality = quality;
        }
    }

    selected_meta
}

/// Returns a sortable quality measure of this object for the given purpose.
///
/// Lower quality number is better.
fn object_quality(meta_lookup: &FoundMeta, purpose: ObjectPurpose) -> u8 {
    match &meta_lookup.handle {
        Ok(object_meta) => match purpose {
            ObjectPurpose::Unwind if object_meta.features.has_unwind_info => 0,
            ObjectPurpose::Debug if object_meta.features.has_debug_info => 0,
            ObjectPurpose::Debug if object_meta.features.has_symbols => 1,
            ObjectPurpose::Source if object_meta.features.has_sources => 0,
            _ => 2,
        },
        Err(_) => 3,
    }
}

/// Whether the object provides the required features for the given purpose.
fn object_has_features(meta_handle: &ObjectMetaHandle, purpose: ObjectPurpose) -> bool {
    match purpose {
        ObjectPurpose::Unwind => meta_handle.features.has_unwind_info,
        ObjectPurpose::Debug => {
            meta_handle.features.has_debug_info || meta_handle.features.has_symbols
        }
        ObjectPurpose::Source => meta_handle.features.has_sources,
    }
}

/// Creates collection of all the DIF object candidates used in the metadata lookups.
///
/// If there were any sources which did not return any [`DownloadService::list_files`]
/// results they will get a [`ObjectDownloadInfo::NotFound`] entry with a location of `*`.
/// In practice this will only affect the `sentry` source for now as all other sources
/// always return [`DownloadService::list_files`] results.
fn create_candidates(sources: &[SourceConfig], lookups: &[FoundMeta]) -> AllObjectCandidates {
    let mut source_ids: BTreeSet<SourceId> =
        sources.iter().map(|source| source.id()).cloned().collect();
    let mut candidates: Vec<ObjectCandidate> = Vec::with_capacity(lookups.len() + source_ids.len());

    for meta_lookup in lookups.iter() {
        let source_id = meta_lookup.file_source.source_id();
        source_ids.take(source_id);
        candidates.push(create_candidate_info(meta_lookup));
    }

    // Create a NotFound entry for each source from which we did not try and fetch anything.
    for source_id in source_ids {
        let info = ObjectCandidate {
            source: source_id,
            location: RemoteFileUri::new("No object files listed on this source"),
            download: ObjectDownloadInfo::NotFound,
            unwind: Default::default(),
            debug: Default::default(),
        };
        candidates.push(info);
    }

    candidates.into()
}

/// Build the [`ObjectCandidate`] info for the provided meta lookup result.
fn create_candidate_info(meta_lookup: &FoundMeta) -> ObjectCandidate {
    let source = meta_lookup.file_source.source_id().clone();
    let location = meta_lookup.file_source.uri();
    let download = match &meta_lookup.handle {
        Ok(handle) => ObjectDownloadInfo::Ok {
            features: handle.features(),
        },
        Err(error) => match error {
            CacheError::NotFound => ObjectDownloadInfo::NotFound,
            CacheError::PermissionDenied(msg) => ObjectDownloadInfo::NoPerm {
                details: msg.clone(),
            },
            CacheError::Malformed(_) => ObjectDownloadInfo::Malformed,
            err => ObjectDownloadInfo::Error {
                details: err.to_string(),
            },
        },
    };

    ObjectCandidate {
        source,
        location,
        download,
        unwind: Default::default(),
        debug: Default::default(),
    }
}
