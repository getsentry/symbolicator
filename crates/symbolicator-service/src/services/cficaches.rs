use std::fs::File;
use std::io::{self, BufWriter};
use std::sync::Arc;
use std::time::Duration;

use futures::future::BoxFuture;
use minidump_unwind::SymbolFile;
use tempfile::NamedTempFile;

use symbolic::cfi::CfiCache;
use symbolic::common::ByteView;
use symbolicator_sources::{FileType, ObjectId, ObjectType, SourceConfig};

use crate::caching::{
    Cache, CacheEntry, CacheError, CacheItemRequest, CacheVersions, Cacher, SharedCacheRef,
};
use crate::services::objects::{
    FindObject, ObjectHandle, ObjectMetaHandle, ObjectPurpose, ObjectsActor,
};
use crate::types::{CandidateStatus, Scope};
use crate::utils::futures::{m, measure};
use crate::utils::sentry::ConfigureScope;

use super::caches::versions::CFICACHE_VERSIONS;
use super::derived::{derive_from_object_handle, DerivedCache};

type CfiItem = (u32, Option<Arc<SymbolFile>>);

#[tracing::instrument(skip_all)]
fn parse_cfi_cache(bytes: ByteView<'static>) -> CacheEntry<CfiItem> {
    let weight = bytes.len().try_into().unwrap_or(u32::MAX);
    // NOTE: we estimate the in-memory structures to be ~8x as heavy in memory as on disk
    let weight = weight.saturating_mul(8);

    let cfi_cache = CfiCache::from_bytes(bytes).map_err(CacheError::from_std_error)?;

    if cfi_cache.as_slice().is_empty() {
        return Ok((weight, None));
    }

    let symbol_file =
        SymbolFile::from_bytes(cfi_cache.as_slice()).map_err(CacheError::from_std_error)?;

    Ok((weight, Some(Arc::new(symbol_file))))
}

#[derive(Clone, Debug)]
pub struct CfiCacheActor {
    cficaches: Arc<Cacher<FetchCfiCacheInternal>>,
    objects: ObjectsActor,
}

impl CfiCacheActor {
    pub fn new(cache: Cache, shared_cache: SharedCacheRef, objects: ObjectsActor) -> Self {
        CfiCacheActor {
            cficaches: Arc::new(Cacher::new(cache, shared_cache)),
            objects,
        }
    }
}

#[derive(Clone, Debug)]
struct FetchCfiCacheInternal {
    objects_actor: ObjectsActor,
    meta_handle: Arc<ObjectMetaHandle>,
}

/// Extracts the Call Frame Information (CFI) from an object file.
///
/// The extracted CFI is written to `path` in symbolic's
/// [`CfiCache`] format.
#[tracing::instrument(skip_all)]
async fn compute_cficache(
    objects_actor: &ObjectsActor,
    meta_handle: Arc<ObjectMetaHandle>,
    temp_file: &mut NamedTempFile,
) -> CacheEntry {
    let object = objects_actor.fetch(meta_handle).await?;

    write_cficache(temp_file.as_file_mut(), &object)
}

impl CacheItemRequest for FetchCfiCacheInternal {
    type Item = CfiItem;

    const VERSIONS: CacheVersions = CFICACHE_VERSIONS;

    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheEntry> {
        let future = compute_cficache(&self.objects_actor, self.meta_handle.clone(), temp_file);

        let timeout = Duration::from_secs(1200);
        let future = tokio::time::timeout(timeout, future);
        let future = measure("cficaches", m::timed_result, future);
        Box::pin(async move { future.await.map_err(|_| CacheError::Timeout(timeout))? })
    }

    fn load(&self, data: ByteView<'static>) -> CacheEntry<Self::Item> {
        parse_cfi_cache(data)
    }

    fn weight(item: &Self::Item) -> u32 {
        item.0.max(std::mem::size_of::<Self::Item>() as u32)
    }
}

/// Information for fetching the symbols for this cficache
#[derive(Debug, Clone)]
pub struct FetchCfiCache {
    pub object_type: ObjectType,
    pub identifier: ObjectId,
    pub sources: Arc<[SourceConfig]>,
    pub scope: Scope,
}

pub type FetchedCfiCache = DerivedCache<Option<Arc<SymbolFile>>>;

impl CfiCacheActor {
    /// Fetches the CFI cache file for a given code module.
    ///
    /// The code object can be identified by a combination of the code-id, debug-id and
    /// debug filename (the basename).  To do this it looks in the existing cache with the
    /// given scope and if it does not yet exist in cached form will fetch the required DIFs
    /// and compute the required CFI cache file.
    pub async fn fetch(&self, request: FetchCfiCache) -> FetchedCfiCache {
        let found_object = self
            .objects
            .find(FindObject {
                filetypes: FileType::from_object_type(request.object_type),
                identifier: request.identifier,
                sources: request.sources,
                scope: request.scope,
                purpose: ObjectPurpose::Unwind,
            })
            .await;

        derive_from_object_handle(found_object, CandidateStatus::Unwind, |meta_handle| {
            let cache_key = meta_handle.cache_key();
            let request = FetchCfiCacheInternal {
                objects_actor: self.objects.clone(),
                meta_handle,
            };
            async {
                let entry = self.cficaches.compute_memoized(request, cache_key).await;

                entry.map(|item| item.1)
            }
        })
        .await
    }
}

/// Extracts the CFI from an object file, writing it to a CFI file.
///
/// The source file is probably an executable or so, the resulting file is in the format of
/// [`CfiCache`].
#[tracing::instrument(skip_all)]
fn write_cficache(file: &mut File, object_handle: &ObjectHandle) -> CacheEntry {
    object_handle.configure_scope();

    tracing::debug!("Converting cficache for {}", object_handle.cache_key);

    let cficache = CfiCache::from_object(object_handle.object()).map_err(|e| {
        let dynerr: &dyn std::error::Error = &e; // tracing expects a `&dyn Error`
        tracing::error!(error = dynerr, "Could not process CFI Cache");

        CacheError::Malformed(e.to_string())
    })?;

    let mut writer = BufWriter::new(file);
    cficache.write_to(&mut writer)?;

    let file = writer.into_inner().map_err(io::Error::from)?;
    file.sync_all()?;

    Ok(())
}
