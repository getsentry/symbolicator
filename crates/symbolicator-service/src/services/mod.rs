//! Provides the internal shared Symbolicator services and a way to initialize them.
//!
//! Symbolicator operates a number of independent services defined in this module for downloading,
//! cache management, and file access.
//! [`SharedServices`] initializes all these internal services according to the provided [`Config`].
//!
//! The internal services require a separate asynchronous runtimes dedicated for I/O-intensive work,
//! such as downloads and access to the shared cache.

use std::sync::Arc;

use anyhow::{Context, Result};

use crate::caching::{Caches, SharedCacheRef, SharedCacheService};
use crate::config::Config;

pub mod bitcode;
pub mod caches;
pub mod cficaches;
pub mod derived;
pub mod download;
mod fetch_file;
pub mod il2cpp;
mod minidump;
mod module_lookup;
pub mod objects;
pub mod ppdb_caches;
pub mod symbolication;
pub mod symcaches;

use self::caches::SourceFilesCache;
use self::download::DownloadService;
use self::objects::ObjectsActor;

pub use self::symbolication::ScrapingConfig;
pub use fetch_file::fetch_file;

pub struct SharedServices {
    pub config: Config,
    pub caches: Caches,
    pub download_svc: Arc<DownloadService>,
    pub shared_cache: SharedCacheRef,
    pub objects: ObjectsActor,
    pub sourcefiles_cache: Arc<SourceFilesCache>,
}

impl SharedServices {
    pub fn new(config: Config, io_pool: tokio::runtime::Handle) -> Result<Self> {
        let caches = Caches::from_config(&config).context("failed to create local caches")?;
        caches
            .clear_tmp(&config)
            .context("failed to clear tmp caches")?;

        let download_svc = DownloadService::new(&config, io_pool.clone());

        let shared_cache = SharedCacheService::new(config.shared_cache.clone(), io_pool);

        let sourcefiles_cache = Arc::new(SourceFilesCache::new(
            caches.sourcefiles.clone(),
            shared_cache.clone(),
            download_svc.clone(),
        ));

        let objects = ObjectsActor::new(
            caches.object_meta.clone(),
            caches.objects.clone(),
            shared_cache.clone(),
            download_svc.clone(),
        );

        Ok(Self {
            config,
            caches,
            download_svc,
            shared_cache,
            objects,
            sourcefiles_cache,
        })
    }
}
