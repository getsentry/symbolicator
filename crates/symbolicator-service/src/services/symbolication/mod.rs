use symbolicator_sources::ObjectId;

use crate::services::cficaches::{CfiCacheActor, CfiCacheError};
use crate::services::objects::ObjectsActor;
use crate::services::ppdb_caches::{PortablePdbCacheActor, PortablePdbCacheError};
use crate::services::symcaches::{SymCacheActor, SymCacheError};
use crate::types::{ObjectFileStatus, RawObjectInfo};

mod apple;
mod module_lookup;
mod process_minidump;
// we should really rename this here to the `SymbolicatorService`, as it does a lot more
// than just symbolication ;-)
#[allow(clippy::module_inception)]
mod symbolication;

pub use symbolication::{StacktraceOrigin, SymbolicateStacktraces};

#[derive(Clone, Debug)]
pub struct SymbolicationActor {
    objects: ObjectsActor,
    symcaches: SymCacheActor,
    cficaches: CfiCacheActor,
    ppdb_caches: PortablePdbCacheActor,
    diagnostics_cache: crate::cache::Cache,
}

impl SymbolicationActor {
    pub fn new(
        objects: ObjectsActor,
        symcaches: SymCacheActor,
        cficaches: CfiCacheActor,
        ppdb_caches: PortablePdbCacheActor,
        diagnostics_cache: crate::cache::Cache,
    ) -> Self {
        SymbolicationActor {
            objects,
            symcaches,
            cficaches,
            ppdb_caches,
            diagnostics_cache,
        }
    }
}

impl From<&CfiCacheError> for ObjectFileStatus {
    fn from(e: &CfiCacheError) -> ObjectFileStatus {
        match e {
            CfiCacheError::Fetching(_) => ObjectFileStatus::FetchingFailed,
            // nb: Timeouts during download are also caught by Fetching
            CfiCacheError::Timeout => ObjectFileStatus::Timeout,
            CfiCacheError::ObjectParsing(_) => ObjectFileStatus::Malformed,

            _ => {
                // Just in case we didn't handle an error properly,
                // capture it here. If an error was captured with
                // `capture_error` further down in the callstack, it
                // should be explicitly handled here as a
                // SymCacheError variant.
                sentry::capture_error(e);
                ObjectFileStatus::Other
            }
        }
    }
}

impl From<&SymCacheError> for ObjectFileStatus {
    fn from(e: &SymCacheError) -> ObjectFileStatus {
        match e {
            SymCacheError::Fetching(_) => ObjectFileStatus::FetchingFailed,
            // nb: Timeouts during download are also caught by Fetching
            SymCacheError::Timeout => ObjectFileStatus::Timeout,
            SymCacheError::Malformed => ObjectFileStatus::Malformed,
            SymCacheError::ObjectParsing(_) => ObjectFileStatus::Malformed,
            _ => {
                // Just in case we didn't handle an error properly,
                // capture it here. If an error was captured with
                // `capture_error` further down in the callstack, it
                // should be explicitly handled here as a
                // SymCacheError variant.
                sentry::capture_error(e);
                ObjectFileStatus::Other
            }
        }
    }
}

impl From<&PortablePdbCacheError> for ObjectFileStatus {
    fn from(e: &PortablePdbCacheError) -> ObjectFileStatus {
        match e {
            PortablePdbCacheError::Fetching(_) => ObjectFileStatus::FetchingFailed,
            // nb: Timeouts during download are also caught by Fetching
            PortablePdbCacheError::Timeout => ObjectFileStatus::Timeout,
            PortablePdbCacheError::Malformed => ObjectFileStatus::Malformed,
            PortablePdbCacheError::PortablePdbParsing(_) => ObjectFileStatus::Malformed,
            _ => {
                // Just in case we didn't handle an error properly,
                // capture it here. If an error was captured with
                // `capture_error` further down in the callstack, it
                // should be explicitly handled here as a
                // PortablePdbCacheError variant.
                sentry::capture_error(e);
                ObjectFileStatus::Other
            }
        }
    }
}

fn object_id_from_object_info(object_info: &RawObjectInfo) -> ObjectId {
    ObjectId {
        debug_id: match object_info.debug_id.as_deref() {
            None | Some("") => None,
            Some(string) => string.parse().ok(),
        },
        code_id: match object_info.code_id.as_deref() {
            None | Some("") => None,
            Some(string) => string.parse().ok(),
        },
        debug_file: object_info.debug_file.clone(),
        code_file: object_info.code_file.clone(),
        object_type: object_info.ty,
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;

    use symbolicator_sources::{ObjectType, SourceConfig};

    use crate::config::Config;
    use crate::services::create_service;
    use crate::services::symbolication::module_lookup::ModuleLookup;
    use crate::test::{self, fixture};
    use crate::types::{CompleteObjectInfo, RawFrame, RawStacktrace, Scope};
    use crate::utils::addr::AddrMode;
    use crate::utils::hex::HexValue;

    /// Setup tests and create a test service.
    ///
    /// This function returns a tuple containing the service to test, and a temporary cache
    /// directory. The directory is cleaned up when the [`TempDir`] instance is dropped. Keep it as
    /// guard until the test has finished.
    ///
    /// The service is configured with `connect_to_reserved_ips = True`. This allows to use a local
    /// symbol server to test object file downloads.
    pub(crate) async fn setup_service() -> (SymbolicationActor, test::TempDir) {
        test::setup();

        let cache_dir = test::tempdir();

        let config = Config {
            cache_dir: Some(cache_dir.path().to_owned()),
            connect_to_reserved_ips: true,
            ..Default::default()
        };
        let handle = tokio::runtime::Handle::current();
        let (symbolication, _objects) = create_service(&config, handle).await.unwrap();

        (symbolication, cache_dir)
    }

    fn get_symbolication_request(sources: Vec<SourceConfig>) -> SymbolicateStacktraces {
        SymbolicateStacktraces {
            scope: Scope::Global,
            signal: None,
            sources: Arc::from(sources),
            origin: StacktraceOrigin::Symbolicate,
            stacktraces: vec![RawStacktrace {
                frames: vec![RawFrame {
                    instruction_addr: HexValue(0x1_0000_0fa0),
                    ..RawFrame::default()
                }],
                ..RawStacktrace::default()
            }],
            modules: vec![CompleteObjectInfo::from(RawObjectInfo {
                ty: ObjectType::Macho,
                code_id: Some("502fc0a51ec13e479998684fa139dca7".to_owned().to_lowercase()),
                debug_id: Some("502fc0a5-1ec1-3e47-9998-684fa139dca7".to_owned()),
                image_addr: HexValue(0x1_0000_0000),
                image_size: Some(4096),
                code_file: None,
                debug_file: None,
                checksum: None,
            })],
        }
    }

    /// Helper to redact the port number from localhost URIs in insta snapshots.
    ///
    /// Since we use a localhost source on a random port during tests we get random port
    /// numbers in URI of the dif object file candidates.  This redaction masks this out.
    pub(crate) fn redact_localhost_port(
        value: insta::internals::Content,
        _path: insta::internals::ContentPath<'_>,
    ) -> impl Into<insta::internals::Content> {
        let re = regex::Regex::new(r"^http://localhost:[0-9]+").unwrap();
        re.replace(value.as_str().unwrap(), "http://localhost:<port>")
            .into_owned()
    }

    macro_rules! assert_snapshot {
        ($e:expr) => {
            ::insta::assert_yaml_snapshot!($e, {
                ".**.location" => ::insta::dynamic_redaction(
                    $crate::services::symbolication::tests::redact_localhost_port
                )
            });
        }
    }

    #[tokio::test]
    async fn test_remove_bucket() {
        // Test with sources first, and then without. This test should verify that we do not leak
        // cached debug files to requests that no longer specify a source.

        let (symbolication, _cache_dir) = setup_service().await;
        let (_symsrv, source) = test::symbol_server();

        let request = get_symbolication_request(vec![source]);
        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());

        let request = get_symbolication_request(vec![]);
        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());
    }

    #[tokio::test]
    async fn test_add_bucket() {
        // Test without sources first, then with. This test should verify that we apply a new source
        // to requests immediately.

        let (symbolication, _cache_dir) = setup_service().await;
        let (_symsrv, source) = test::symbol_server();

        let request = get_symbolication_request(vec![]);
        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());

        let request = get_symbolication_request(vec![source]);
        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());
    }

    #[tokio::test]
    async fn test_apple_crash_report() {
        let (symbolication, _cache_dir) = setup_service().await;
        let (_symsrv, source) = test::symbol_server();

        let report_file = std::fs::File::open(fixture("apple_crash_report.txt")).unwrap();

        let response = symbolication
            .process_apple_crash_report(Scope::Global, report_file, Arc::new([source]))
            .await;

        assert_snapshot!(response.unwrap());
    }

    #[tokio::test]
    async fn test_wasm_payload() {
        let (symbolication, _cache_dir) = setup_service().await;
        let (_symsrv, source) = test::symbol_server();

        let modules: Vec<RawObjectInfo> = serde_json::from_str(
            r#"[
              {
                "type":"wasm",
                "debug_id":"bda18fd8-5d4a-4eb8-9302-2d6bfad846b1",
                "code_id":"bda18fd85d4a4eb893022d6bfad846b1",
                "debug_file":"file://foo.invalid/demo.wasm"
              }
            ]"#,
        )
        .unwrap();

        let stacktraces = serde_json::from_str(
            r#"[
              {
                "frames":[
                  {
                    "instruction_addr":"0x8c",
                    "addr_mode":"rel:0"
                  }
                ]
              }
            ]"#,
        )
        .unwrap();

        let request = SymbolicateStacktraces {
            modules: modules.into_iter().map(From::from).collect(),
            stacktraces,
            signal: None,
            origin: StacktraceOrigin::Symbolicate,
            sources: Arc::new([source]),
            scope: Default::default(),
        };

        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());
    }

    #[tokio::test]
    async fn test_source_candidates() {
        let (symbolication, _cache_dir) = setup_service().await;
        let (_symsrv, source) = test::symbol_server();

        // its not wasm, but that is the easiest to write tests because of relative
        // addressing ;-)
        let modules: Vec<RawObjectInfo> = serde_json::from_str(
            r#"[
              {
                "type":"wasm",
                "debug_id":"7f883fcd-c553-36d0-a809-b0150f09500b",
                "code_id":"7f883fcdc55336d0a809b0150f09500b"
              }
            ]"#,
        )
        .unwrap();

        let stacktraces = serde_json::from_str(
            r#"[
              {
                "frames":[
                  {
                    "instruction_addr":"0x3880",
                    "addr_mode":"rel:0"
                  }
                ]
              }
            ]"#,
        )
        .unwrap();

        let request = SymbolicateStacktraces {
            modules: modules.into_iter().map(From::from).collect(),
            stacktraces,
            signal: None,
            origin: StacktraceOrigin::Symbolicate,
            sources: Arc::new([source]),
            scope: Default::default(),
        };

        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());
    }

    #[test]
    fn test_symcache_lookup_open_end_addr() {
        test::setup();

        // The Rust SDK and some other clients sometimes send zero-sized images when no end addr
        // could be determined. Symbolicator should still resolve such images.
        let info = CompleteObjectInfo::from(RawObjectInfo {
            ty: ObjectType::Unknown,
            code_id: None,
            debug_id: None,
            code_file: None,
            debug_file: None,
            image_addr: HexValue(42),
            image_size: Some(0),
            checksum: None,
        });

        let lookup = ModuleLookup::new(Scope::Global, Arc::new([]), std::iter::once(info.clone()));

        let lookup_result = lookup.lookup_cache(43, AddrMode::Abs).unwrap();
        assert_eq!(lookup_result.module_index, 0);
        assert_eq!(lookup_result.object_info, &info);
        assert!(lookup_result.cache.is_none());
    }

    #[tokio::test]
    async fn test_dotnet_integration() {
        let (symbolication, _cache_dir) = setup_service().await;
        let (_srv, source) = test::symbol_server();

        let modules: Vec<RawObjectInfo> = serde_json::from_str(
            r#"[
              {
                "type":"pe_dotnet",
                "debug_file":"integration.pdb",
                "debug_id":"0c1033f78632492e91c6c314b72e1920e60b819d"
              }
            ]"#,
        )
        .unwrap();

        let stacktraces = serde_json::from_str(
            r#"[
              {
                "frames":[
                  {
                    "instruction_addr": 10,
                    "function_id": 6,
                    "addr_mode":"rel:0"
                  },
                  {
                    "instruction_addr": 6,
                    "function_id": 5,
                    "addr_mode": "rel:0"
                  },
                  {
                    "instruction_addr": 0,
                    "function_id": 3,
                    "addr_mode": "rel:0"
                  },
                  {
                    "instruction_addr": 0,
                    "function_id": 2,
                    "addr_mode": "rel:0"
                  },
                  {
                    "instruction_addr": 45,
                    "function_id": 1,
                    "addr_mode": "rel:0"
                  }
                ]
              }
            ]"#,
        )
        .unwrap();

        let request = SymbolicateStacktraces {
            modules: modules.into_iter().map(From::from).collect(),
            stacktraces,
            signal: None,
            origin: StacktraceOrigin::Symbolicate,
            sources: Arc::new([source]),
            scope: Default::default(),
        };

        let response = symbolication.symbolicate(request).await;

        assert_snapshot!(response.unwrap());
    }
}
