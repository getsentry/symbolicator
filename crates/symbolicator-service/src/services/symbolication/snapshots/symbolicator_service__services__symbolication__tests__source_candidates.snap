---
source: crates/symbolicator-service/src/services/symbolication/mod.rs
assertion_line: 409
expression: response.unwrap()
---
stacktraces:
  - frames:
      - status: symbolicated
        original_index: 0
        addr_mode: "rel:0"
        instruction_addr: "0x3880"
        lang: c
        symbol: trigger_crash
        sym_addr: "0x3860"
        function: trigger_crash
        filename: example.c
        abs_path: /Users/swatinem/Coding/sentry-native/examples/example.c
        lineno: 60
        pre_context:
          - ""
          - static void
          - trigger_crash()
          - "{"
        context_line: "    memset((char *)invalid_mem, 1, 100);"
        post_context:
          - "}"
          - ""
          - int
          - "main(int argc, char **argv)"
          - "{"
modules:
  - debug_status: found
    features:
      has_debug_info: true
      has_unwind_info: false
      has_symbols: true
      has_sources: true
    arch: x86_64
    type: wasm
    code_id: 7f883fcdc55336d0a809b0150f09500b
    debug_id: 7f883fcd-c553-36d0-a809-b0150f09500b
    image_addr: "0x0"
    candidates:
      - source: local
        location: "http://localhost:<port>/download/7f/883fcdc55336d0a809b0150f09500b"
        download:
          status: notfound
      - source: local
        location: "http://localhost:<port>/download/7f/883fcdc55336d0a809b0150f09500b.debug"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: false
            has_symbols: true
            has_sources: false
        debug:
          status: ok
      - source: local
        location: "http://localhost:<port>/download/7f/883fcdc55336d0a809b0150f09500b.src.zip"
        download:
          status: ok
          features:
            has_debug_info: false
            has_unwind_info: false
            has_symbols: false
            has_sources: true

